### PR TITLE
feat(biome_css_analyze): implement noIrregularWhitespaceCss

### LINKED ISSUE #3264: 📎 Implement `stylelint/no-irregular-whitespace`
## Description

Implement [no-irregular-whitespace](https://stylelint.io/user-guide/rules/no-irregular-whitespace/)

> [!IMPORTANT]
>   - Please skip implementing options for now since we will evaluate users actually want them later.
>   - Please ignore handling extended CSS language cases such as sass and less.
>   - Please skip custom function since we haven't syntax model yet.

**Want to contribute?** Lets you know you are interested! We will assign you to the issue to prevent several people to work on the same issue. Don't worry, we can unassign you later if you are no longer interested in the issue! Read our [contributing guide](https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md) and [analyzer contributing guide](https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by DerTimonius**
## Summary

This PR supercedes #3362 (there were some collisions between the css and the js rule and it was harder to fix than anticipated, so I started over) and implements the `no-irregular-whitespace` rule from [stylelint](https://stylelint.io/user-guide/rules/no-irregular-whitespace/). 

My original implementation only checked for selectors, now this rule checks for complete rule blocks, regardless of type of rule (import rules, container rules, simple rule, you name it).

I was able to reuse most of the logic used in #3333 with some changes.

## Test Plan

I tested some different @-rules, used the whitespace in the selector and in the rule itself. I also used all characters that are declared irregular.

closes #3264

### CODE REVIEW

File: crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    /// ### Valid
+    ///
+    /// ```css
+    /// .firstClass .secondClass {
+    ///   color: red;
+    /// }
+    /// ```
+    ///
+    pub NoIrregularWhitespaceCss {
+        version: "1.8.0",
```
Comment (togami2864) [inline crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs:43]:
```suggestion
        version: "next",
```

File: crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        name: "noIrregularWhitespaceCss",
+        language: "css",
+        recommended: false,
```
Comment (togami2864) [inline crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs:46]:
```suggestion
        recommended: true,
```

File: crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        get_irregular_whitespace(node)
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, range: &Self::State) -> Option<RuleDiagnostic> {
+        //
+        // Read our guidelines to write great diagnostics:
+        // https://docs.rs/biome_analyze/latest/biome_analyze/#what-a-rule-should-say-to-the-user
+        //
```
Comment (togami2864) [inline crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs:66]: Please remove the comment

File: crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    /// ```
+    /// ### Valid
+    ///
+    /// ```css
+    /// .firstClass .secondClass {
+    ///   color: red;
+    /// }
+    /// ```
+    ///
+    pub NoIrregularWhitespaceCss {
```
Comment (togami2864) [inline crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs:42]: `CSS` isn't necessary. One rule can use the same name in multiple languages in Biome.

```suggestion
    pub NoIrregularWhitespace {
```

File: crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs (hunk 2 above)
Comment (ematipico) [reply crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs:46]: I don't think we should recommend it. It's a slow rule, and a niche one.

File: crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs (hunk 4 above)
Comment (DerTimonius) [reply crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs:42]: when trying to rename this, I run into this issue (which is the reason why I opened a new PR for this):
image

Comment (ematipico) [reply crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs:42]: No, problem. Keep the old name and I will handle that once this rule is merged

File: crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+};
+use biome_console::markup;
+use biome_css_syntax::{AnyCssRule, CssLanguage};
+use biome_rowan::{AstNode, Direction, SyntaxToken, TextRange};
+
+const IRREGULAR_WHITESPACES: &[char; 22] = &[
+    '\u{c}', '\u{b}', '\u{85}', '\u{feff}', '\u{a0}', '\u{1680}', '\u{180e}', '\u{2000}',
+    '\u{2001}', '\u{2002}', '\u{2003}', '\u{2004}', '\u{2005}', '\u{2006}', '\u{2007}', '\u{2008}',
+    '\u{2009}', '\u{200a}', '\u{200b}', '\u{202f}', '\u{205f}', '\u{3000}',
+];
```
Comment (Mouvedia) [inline crates/biome_css_analyze/src/lint/nursery/no_irregular_whitespace_css.rs:12]: question

Are `\u{2028}` and `\u{2029}` missing?

see w3c/csswg-drafts#6992

**Comment by ematipico** [comment]
@DerTimonius are you still interested in this PR?

**Comment by DerTimonius** [comment]
@ematipico yes, I was wondering what is still missing

**Comment by ematipico** [comment]
Just fix the conflicts and I'll merge it

**Comment by DerTimonius** [comment]
@ematipico thanks, resolved the conflicts 🙂

**Pull request merged**